src/ui/share-dialog.vala

```diff
@@ -0,0 +1,123 @@
+
+using Gtk;
+using Soup;
+
+namespace Peek.Ui {
+
+  [GtkTemplate (ui = "/com/uploadedlobster/peek/share-dialog.ui")]
+  class ShareDialog : Window {
+  private static Gtk.Window? instance;
+
+    public static Gtk.Window present_single_instance (Gtk.Window main_window) {
+      if (instance == null) {
+        instance = new ShareDialog ();
+        instance.delete_event.connect ((event) => {
+          instance = null;
+          main_window.set_keep_above (true);
+          return false;
+        });
+      }
+
+      instance.transient_for = main_window;
+      main_window.set_keep_above (false);
+      instance.present ();
+      return instance;
+    }
+
+    public static string file { get; set; }
+    public static string file_type { get; set; }
+
+    public static void filename (string out_file) {
+        file=out_file;
+	}
+
+    public static void get_file_ext (string file_ext) {
+        file_type=file_ext;
+	}
+
+    [GtkChild]
+    private Gtk.Image check_1;
+
+    [GtkChild]
+    private Gtk.Image check_2;
+
+    [GtkChild]
+    private Gtk.ListBox options_list;
+
+    [GtkChild]
+    private Gtk.ListBoxRow row_1;
+
+    [GtkChild]
+    private Gtk.ListBoxRow row_2;
+
+    [GtkCallback]
+    private void on_options_list_row_selected () {
+      var selection = options_list.get_selected_row ();
+      if(file_type == "webm" || file_type == "mp4"){
+          row_1.set_selectable(false);
+          check_2.show();
+          check_1.hide();
+          options_list.select_row(row_2);
+          debug("WebM and Mp4 files are not supported with in imgur upload api");
+      }else {
+          row_1.set_selectable(true);
+          check_1.show();
+          if (selection == row_1) {
+            check_2.hide();
+         }else if (selection == row_2 ){
+            check_2.show();
+            check_1.hide();
+            }
+      }
+
+
+    }
+    [GtkCallback]
+    private void on_confirm_option_clicked() {
+      if (options_list.get_selected_row () == row_1) {
+        string image_uri = file;
+        string image=image_uri.replace("%20"," ");
+        debug(image);
+        uint8[] data;
+
+        try {
+            GLib.FileUtils.get_data(image.split("://")[1], out data);
```

Comment:
If you use `File.get_path()` instead of `File.get_uri()` you can avoid this string manipulation. Also the `%20` replacement above and likely some other url encoded things that could sneak into the URI part.